src/app/CASEClientPool.h

```diff
@@ -0,0 +1,36 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/CASEClient.h>
+#include <lib/support/Pool.h>
+
+namespace chip {
+
+class CASEClientPool
```

Comment:
I understand that the session manager needs to allocate clients for inbound session establishment.  But for outbound session establishment, while this may be expedient for constrained targets, this is going to be quite awkward for larger nodes.

If I were asked how many clients I would like for outbound session establishment, I would probably say something like SIZE_MAX.  As in, the sdk shouldn't impose any particular limitation.

It appears it may be possible to use a different client pool than the one inside the session manager when establishing outbound sessions.  Can we confirm that this is true?  That's a start, but it's not enough.

We also need to make sure that implementations can use a more suitable implementation for outbound connections.  Can we make the client pool an abstract interface?  The operational device proxy shouldn't be restricted to using this particular concrete pool implementation with only 2 slots, or any fixed number of slots at all.

As to this concrete implementation of the client pool, I understand that it will exist and, again, the session manager needs its own pool for inbound session establishment.  But even for that case, this is too restrictive.  Can we make this templated on N, the size of the pool?

Reply:
Inbound session estatlishment won't be affected.
Having a client pool with 2 slots doesn't mean the we can only have 2 operational device proxy at the same time. The client is only used in the 3-step Sigma handshaking. After that the `CASESession` is no longer required for an operational device. The size of the client pool only limits the number of simultaneous sigma handshakes.
Making `ClientPool` a template or a abstract interface is possible but a simple implementation may suffice for both devices and POSIX controllers. Maybe the key problem is that the `CASSessionManager` is no longer a proper name since operational devices are no longer bound to a CASESession?

Reply:
> Inbound session estatlishment won't be affected.

Thanks for the clarification.

> Having a client pool with 2 slots doesn't mean the we can only have 2 operational device proxy at the same time. The client is only used in the 3-step Sigma handshaking. 

But a non-constrained controller booting up and establishing sessions with a large number of peers all at once is still a very likely use case.  Serializing session establishment with a large number of peers could be extremely time consuming.

With this PR, is there still a straight forward way to bypass the client pool and allocate my own clients, or otherwise perform session establishment with a large number of peers concurrently?